File: src/main/java/org/distril/beengine/network/Network.java
package org.distril.beengine.network;

import com.nukkitx.protocol.bedrock.*;
import com.nukkitx.protocol.bedrock.v567.Bedrock_v567patch;
import lombok.extern.log4j.Log4j2;
import org.distril.beengine.player.handler.LoginPacketHandler;
import org.distril.beengine.server.Server;

import java.net.InetSocketAddress;
import java.util.concurrent.CompletionException;

@Log4j2
public class Network implements BedrockServerEventHandler {

	public static final BedrockPacketCodec CODEC = Bedrock_v567patch.BEDROCK_V567PATCH;

	public static final BedrockPong PONG = new BedrockPong();

	static {
		PONG.setEdition("MCPE");
		PONG.setGameType("Survival");
		PONG.setProtocolVersion(CODEC.getProtocolVersion());
		PONG.setVersion(CODEC.getMinecraftVersion());
		PONG.setPlayerCount(0);
	}

	private final Server server;

	private final BedrockServer bedrockServer;

	public Network(Server server, String ip, int port) {
		this.server = server;

		var bindAddress = new InetSocketAddress(ip, port);
		this.bedrockServer = new BedrockServer(bindAddress, Runtime.getRuntime().availableProcessors());
		this.bedrockServer.setHandler(this);

		PONG.setMotd(server.getSettings().getMotd());
		PONG.setSubMotd(PONG.getMotd());
		PONG.setMaximumPlayerCount(server.getSettings().getMaximumPlayers());
		PONG.setIpv4Port(port);
		PONG.setIpv6Port(port);
	}

	public void start() throws Exception {
		try {
			this.bedrockServer.bind().join();

			var bindAddress = this.bedrockServer.getBindAddress();

			var ip = bindAddress.getHostString();
			var port = bindAddress.getPort();
			log.info("Server started on " + ip + ":" + port);
		} catch (CompletionException exception) {
			if (exception.getCause() instanceof Exception) {
				throw (Exception) exception.getCause();
			}

			throw exception;
		}
	}

	@Override
	public boolean onConnectionRequest(InetSocketAddress address) {
		return true;
	}

	@Override
	public BedrockPong onQuery(InetSocketAddress address) {
		return PONG;
	}

	@Override
	public void onSessionCreation(BedrockServerSession session) {
		session.setCompressionLevel(this.server.getSettings().getCompressionLevel());
		session.setLogging(false);
		session.setPacketCodec(CODEC);
		session.setPacketHandler(new LoginPacketHandler(session, this.server));
	}

	public void stop() {
		this.bedrockServer.close(true);
	}
}

File: src/main/kotlin/org/distril/beengine/network/Network.kt
package org.distril.beengine.network

import com.nukkitx.protocol.bedrock.BedrockPong
import com.nukkitx.protocol.bedrock.BedrockServer
import com.nukkitx.protocol.bedrock.BedrockServerEventHandler
import com.nukkitx.protocol.bedrock.BedrockServerSession
import com.nukkitx.protocol.bedrock.v567.Bedrock_v567patch
import org.distril.beengine.player.handler.LoginPacketHandler
import org.distril.beengine.server.Server
import org.distril.beengine.util.Utils.getLogger
import java.net.InetSocketAddress
import java.util.concurrent.CompletionException

class Network(
    ip: String,
    port: Int
) : BedrockServerEventHandler {

    private val bedrockServer: BedrockServer

    init {
        val bindAddress = InetSocketAddress(ip, port)
        bedrockServer = BedrockServer(bindAddress, Runtime.getRuntime().availableProcessors())
        bedrockServer.handler = this


        PONG.apply {
            motd = Server.settings.motd
            subMotd = motd
            maximumPlayerCount = Server.settings.maximumPlayers
            ipv4Port = port
            ipv6Port = ipv4Port
        }
    }

    fun start() {
        try {
            bedrockServer.bind().join()
            log.info("Server started on ${bedrockServer.bindAddress} with ${CODEC.minecraftVersion} Minecraft version")
        } catch (exception: CompletionException) {
            if (exception.cause is Exception) throw exception.cause as Exception

            throw exception
        }
    }

    override fun onConnectionRequest(address: InetSocketAddress) = true

    override fun onQuery(address: InetSocketAddress) = PONG

    override fun onSessionCreation(session: BedrockServerSession) {
        session.apply {
            compressionLevel = Server.settings.compressionLevel
            isLogging = false
            packetCodec = CODEC
            packetHandler = LoginPacketHandler(session)
        }
    }

    fun stop() = this.bedrockServer.close(true)

    companion object {

        private val log = Network.getLogger()

        val CODEC = Bedrock_v567patch.BEDROCK_V567PATCH!!

        val PONG = BedrockPong().apply {
            edition = "MCPE"
            gameType = "Survival"
            protocolVersion = CODEC.protocolVersion
            version = CODEC.minecraftVersion
            playerCount = 0
        }
    }
}
